M365 Profile Control Files

skandwal-mitre opened the pull request:
Added Initial Control Files to M365 Profile

skandwal-mitre committed:
added initital control files to m365 profile

wdower commented on .DS_Store:
.DS_Store files are OS-specific. Please remove these

Amndeep7 commented on .DS_Store:
delete thsi file

Amndeep7 commented on controls/.DS_Store:
delete this file.  you're going to want to set up a .gitignore to make this automatic.

Amndeep7 commented on controls/microsoft-365-foundations-1.1.1.rb:
```diff
@@ -0,0 +1,34 @@
+        5.When prompted to assign licenses select as needed Microsoft Entra ID P1 or Microsoft Entra ID P2, then click Next.
+        6.Under the Option settings screen you may choose from several types of Administrative access roles. Choose Admin center access followed by the appropriate role then click'
+    impact 0.5
+    tag severity: 'medium '
```
get rid of whitespace like this

aaronlippold commented:
Remove the .DB_Store file from the PR

aaronlippold commented:
You forgot all your `end` statements in the control files. Then get with me and I can help you add all the template .rubocop and other settings file to help clean up your profile.

aaronlippold commented:
Your `inspec.yml` file needs to be updated. You can use https://raw.githubusercontent.com/mitre/aws-foundations-cis-baseline/master/inspec.yml as an example

Amndeep7 commented on controls/microsoft-365-foundations-1.1.1.rb:
```diff
@@ -0,0 +1,34 @@
+
+    describe 'manual' do
+        skip 'manual'
+
```
you're missing `end`s for both the describe block as well as the control as a whole.

Amndeep7 commented on controls/microsoft-365-foundations-1.1.3.rb:
```diff
@@ -0,0 +1,44 @@
+    desc 'More than one global administrator should be designated so a single admin can be monitored and to provide redundancy should a single admin leave an organization. Additionally, there should be no more than four global admins set for any tenant. Ideally global administrators will have no licenses assigned to them.'
+
+    desc 'check'
+    'Ensure that between two and four global admins are designated:
```
not sure if we do multiline strings like this with quotes or the %q syntax - talk with @wdower

aaronlippold replied:
Yes we do, its standard ruby syntax

Amndeep7 commented on controls/microsoft-365-foundations-1.2.2.rb:
```diff
@@ -0,0 +1,46 @@
+        Users with permissions to the group mailbox can send as or send on behalf of the mailbox email address if the administrator has given that user permissions to do that. This is particularly useful for help and support mailboxes because users can send emails from "Contoso Support" or "Building A Reception Desk."
+        Shared mailboxes are created with a corresponding user account using a system generated password that is unknown at the time of creation.
+        The recommended state is Sign in blocked for Shared mailboxes.'
+    
```
you need to make sure to bring across ALL of the metadata defined in the benchmark.  for example, there should be a 'rationale' describe block that includes that section.

Amndeep7 commented on controls/microsoft-365-foundations-1.2.2.rb:
```diff
@@ -0,0 +1,46 @@
+    
+    impact 0.5
+    tag severity: 'medium'
+    
```
probably a good idea to check other profiles to see what we do with cis controls that are 'explicitly not mapped'.  i think we still would want to have that metadata written out here.

Amndeep7 replied:
i think that 'default value' metadata should be passed through as a tag or something?  not sure.  regardless, it needs to be in here somehow as well.

Amndeep7 reviewed (changes requested):
didn't leave comments in every location, but you can apply the comments i left throughout the entire profile.

even if they're all skips, make sure to actually run the profile and load the results into heimdall.  you would've realized that you had syntax issues that way.

skandwal-mitre committed:
Added end statements

skandwal-mitre committed:
Added git ignore/remove ds_store

aaronlippold committed:
Fixed:

- All ruby syntax, formatting, linting and style errors
- Added missing end statments, describe blocks and other inspec style
  isseus
- Fixed escaping issues in check and fix text in multiple controls
- Fixed empty inspec.yml file
- Fixed incorrect LICENSE, NOTEICE and CODE_OF_CONDUCT files
- Ran bundle exec rake lint:auto_correct
- Ran rubocop/cookstyle
- Removed DS_STORE and other unneeded files
- Ran bundle exec inspec . and make sure the profile passed

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
ran linting

Signed-off-by: Aaron Lippold <lippold@gmail.com>

skandwal-mitre committed:
Updates before PTO

skandwal-mitre committed:
Latest changes before PTO

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

sujKandwal committed:
Added rationale, default_value tag, and nist tag

aaronlippold committed:
added link to powershell module install

Signed-off-by: Aaron Lippold <lippold@gmail.com>

skandwal-mitre commented:
Addressed all changes and spoke with Aaron on them. Asked Will and %q syntax is not needed for now.

skandwal-mitre committed:
Automated some controls and designated manual controls

skandwal-mitre committed:
Automated more controls

skandwal-mitre committed:
Added SharePoint controls for automation

wdower commented on README.md:
```diff
@@ -4,6 +4,9 @@
 
 InSpec profile to validate the secure configuration of a Kubernetes node against [DISA's](https://iase.disa.mil/stigs/Pages/index.aspx) Kubernetes Secure Technical Implementation Guide (STIG) Version 1 Release 1.
```
Still referencing the template README -- please update the README to show how to run this against M365

skandwal-mitre committed:
Added pipeline code

aaronlippold committed:
fixing workflow, added missing files, fixed paths, etc.

Signed-off-by: GitHub <noreply@github.com>

wdower commented on .github/workflows/collect_cert.ps1:
We don't need this script any more because you've distilled it down into a one-liner in the test_profile workflow. Please delete

wdower commented on controls/microsoft-365-foundations-1.1.3.rb:
```diff
@@ -0,0 +1,72 @@
+    $client_id = '#{input('client_id')}'
+    $appObjectID = '136a849a-74bc-47ed-9bec-dde306688a6a' #hardly ever used
+    $tenantid = '#{input('tenant_id')}'
+    $clientSecret = '#{input('client_secret')}' #This should not be stored inside of any script; supplied to transmit detail
```
Looks like there's still some of the comments from the original script in here. Not really needed for the InSpec version. Please remove

wdower commented on controls/microsoft-365-foundations-2.1.1.rb:
```diff
@@ -0,0 +1,152 @@
+  end
+
+  policy_names = policy_links_script.stdout.strip.split("\n")
+  policy_names.each do |policy_name|
```
So using a `.each` loop like this works, but is a bit messy -- doing it this way means InSpec prints a bunch of output even for items in the loop that pass the test.

Test output will be cleaner if you try to assemble a list of policies that *fail* the test, then have the `describe` block basically just check if it is empty. An example is here https://github.com/mitre/redhat-enterprise-linux-9-stig-baseline/blob/0d07cf34d8f7c29ae7204f16ae855c52cbdc764b/controls/SV-257966.rb#L59

I don't consider this to be a blocker for the merge but I will add it as an issue to streamline this for v2.

wdower commented on controls/microsoft-365-foundations-2.1.14.rb:
I'm starting to think we might want to take these giant ps1 scripts and put them into the `files` directory at the root of the profile and then just run them in the controls. See https://docs.chef.io/inspec/profiles/files/

This is again a v2 thing since it's not stopping test exection.

wdower commented on controls/microsoft-365-foundations-2.1.6.rb:
```diff
@@ -0,0 +1,101 @@
+    { '7' => ['7.9'] },
+    { '7' => ['7.10'] }
+  ]
+  tag default_value: "BccSuspiciousOutboundAdditionalRecipients : {}
```
I don't think we really need this tag, since default values are defined in `inspec.yml`. It might actually confuse people to leave these in, since it just confused me.

Please strike this tag from all controls.

wdower commented on controls/microsoft-365-foundations-2.1.6.rb:
```diff
@@ -0,0 +1,101 @@
+  powershell_output = JSON.parse(powershell(ensure_exchange_online_spam_policies_set_to_notify_admins_script).stdout.strip)
+  case powershell_output
+  when Hash
+    describe "Ensure the following Exchange Online Spam Policy (#{powershell_output['Name']})" do
```
This is an area where we don't really need to use RSpec syntax and would probably be better served just using classic InSpec syntax:

``` ruby
describe "Ensure the following Exchange Online Spam Policy (#{powershell_output['Name']})" do
    subject { powershell_output['Name'] }
    its('BccSuspiciousOutboundMail') { should cmp true }
    its('NotifyOutboundSpam') { should cmp true }
    . . .
end
```

etc etc

Right now you're basically rewriting the test output as a string when InSpec's default syntax will give you that for free.

wdower commented on m365_results.json:
You shouldn't commit the test results. Remove the file and add a `*.json` to avoid accidentally doing it.

wdower commented on libraries/.gitkeep:
Since we don't have any custom libraries right now, we can delete the `libraries` folder in its entirety.

wdower commented on controls/microsoft-365-foundations-2.4.4.rb:
We can actually just capture the expected list of exclusions as an input. We can assume that exclusions specifically called out in the input are justified, and if we see any that aren't on the list we fail.

Also typo -- 'list of of exclusions'

wdower commented on controls/microsoft-365-foundations-5.2.2.3.rb:
```diff
@@ -0,0 +1,127 @@
+  describe 'Ensure there is no Exchange Online User' do
+    subject { block_basic_policy_data }
+    it 'with AuthenticationPolicy state that is not Block Basic Auth' do
+      failure_message = "Users that failed: #{JSON.pretty_generate(block_basic_policy_data)}"
```
This is a great example of my earlier comment of "write tests that check that the count of failing items is zero"

wdower commented on controls/microsoft-365-foundations-7.2.9.rb:
```diff
@@ -0,0 +1,73 @@
+      expect(subject['ExternalUserExpirationRequired']).to eq(true)
+    end
+    it 'ExternalUserExpireInDays in SharePoint/OneDrive is less than or equal to 30' do
+      expect(subject['ExternalUserExpireInDays']).to be <= 30
```
Let's make an input for that 30. Numeric values like that usually should be parameterized.

wdower commented on inspec.yml:
```diff
@@ -0,0 +1,217 @@
+summary: "InSpec Validation Profile for the CIS Microsoft 365 Foundations Benchmark"
+version: 3.1.0
+inspec_version: ">= 6"
+# supports:
```
Scrub comments in the final product.

wdower commented on inspec.yml:
```diff
@@ -0,0 +1,217 @@
+  value: false
+  required: false
+
+#Controls using this input:
```
These comments are fine. Thanks for keeping track of what input is used in what control.

wdower commented on libraries/.gitkeep:
We can remove this whole folder if there's no actual resource code in it.

wdower reviewed (changes requested):
Great work so far @skandwal-mitre. Some overall notes:

- There are a few styling issues. In several places you are using RSpec syntax for a test when you really don't need to, and should be using basic InSpec syntax to make the tests shorter and clearer. I do not consider this a blocker for merging but we will address it in V2.
- Similarly there are a few places where you stick `describe` blocks inside loops (e.g. `.each`). This is not a good idea for two reasons -- one, you never know when your iterator will not actually have any values to loop over, which stop your test from doing anything, and two because it makes InSpec's output really really long for no good reason since it will print a status for passes and fails when all we really care about are the fails. Again this is not a merge blocker but is a style thing we need to fix.
- I think we need to investigate a way to save off these .ps1 scripts given in the benchmark to a separate directory and then just run them with the controls. This will make them way more readable. I think we can do this by making a `files` directory that InSpec can see inside the controls.
- There's a handful of other direct comments I would like you to address before I call this done.

wdower commented:
Took a look at Heimdall Demo; [9 profile errors](https://heimdall-demo.mitre.org/results/642). Please resolve.

I'm also noticing that a bunch of the actual failing tests don't seem to actually be returning data. Like microsoft-365-foundations-1.3.3:
![image](https://github.com/user-attachments/assets/85d25174-289e-4569-b9bb-4bf2771d9e95)

Not sure if that's correct behavior from the API call or not. Sometimes a fail like that means that your API call isn't actually formed correctly, and the test just interprets it as a fail.

skandwal-mitre committed:
Threshold changes

skandwal-mitre committed:
Remove results

skandwal-mitre committed:
updated pipeline

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Fixing profile errors, removing comments, removing certain files

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
test workflow using control 2.1.5

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Add code to check workflow fix

skandwal-mitre committed:
change workflow to only test 2.1.5 for time being

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
try more code changes to 2.1.5

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow again

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

aaronlippold commented on .github/workflows/test_profile.yml:
```diff
@@ -0,0 +1,96 @@
+        continue-on-error: true
+        run: |
+          bundle exec inspec exec . \
+          --controls=microsoft-365-foundations-2.1.5 \
```
We should figure out how to make this a variable we can pass in via the commit message if we want to be able to do this kind of thing and it should have logic in it to only add the --controls flag if the commit message has that variable

skandwal-mitre committed:
Create devcontainer.json

skandwal-mitre committed:
Check workflow once again with more controls

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 3

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 4

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 5

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 6

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 7

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 8

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 9

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 10

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 11

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 12

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 13

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 14

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 15

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 16

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 17

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 18

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check workflow once again pt 19

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Remove debug prints

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check 1.3.1 failure

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Test 1.3.1 again

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Try fixes on 1.3.1

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Continue looking at fixes

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Try some other control changes

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Test the fixes more

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Fix pipline code

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
adding installs for all automated controls just in case

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Parameterized more and added readme

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check 1.3.1

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
print input values to test

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
test 1.3.1 again

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
test entire profile with fixes

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
test few more things

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
change the test profile

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

wdower committed:
updating README

Signed-off-by: Will <will@dower.dev>

skandwal-mitre committed:
test 5.1.1.1

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
test scope controls

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Test repo cert env variable

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check 5.2.3.4 against input update

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Test whole profile

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Check inputs

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Convert 5.2.3.4 to int

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

skandwal-mitre committed:
Updated data/readme

Signed-off-by: Sujay Kandwal <skandwal@mitre.org>

wdower committed:
Update README.md

typo'd README link
